3주차 미션 제출

jambottle opened the pull request:
# 3주차 미션 제출 (김재원)

## 구현 화면

<img width="1280" alt="week3_02" src="https://user-images.githubusercontent.com/72926450/152649637-b5f6d8fe-b573-4b08-bdda-d16d1b1fa83f.png">

## 접근 방법

* [x] 미션 코드를 구현하기 위해서 어떤 생각들을 했는지 정리해보세요.

**📌 주어진 요구사항을 정확히 숙지하고, 그에 따라 페이지와 컴포넌트를 개발하였습니다.**
- 미션 요구사항에 대한 테스트 코드를 먼저 작성한 뒤, 앞서 작성한 테스트 코드를 바탕으로 페이지 및 컴포넌트 개발
- `Card 컴포넌트`
  - 상품 정보 카드는 별도의 하위 컴포넌트로 분리하여 v-for 디렉티브로 리스트 렌더링 될 수 있도록 구현 (Card.vue)
  - 상품 정보 카드는 상위의 ItemList 페이지로부터 데이터를 prop으로 전달받아 데이터 바인딩 및 리스트 렌더링
- `Header/Nav 영역`
  - 최상단의 Header 영역과 최하단의 Nav 영역, 그리고 그 사이에 2열로 표시될 상품 정보 카드 영역으로 페이지 구분
  - 최상단의 Header 영역은 하단으로 스크롤함에 따라 해당 영역을 숨기는 효과를 조건부 스타일링을 통해 구현

**📌 테스트 코드 리팩토링**
- ItemList 페이지에 대한 테스트 코드로부터 Header/Card/Footer(Nav) 컴포넌트의 테스트 코드를 각각 분리
- ItemList 페이지에 대한 테스트 코드에서는 리스트 렌더링을 확인하기 위해 `findAllComponents` 메서드를 활용
  - 이후, 테스트 코드 내에서 for 반복문을 활용해 컴포넌트별로 포함해야 할 요소들이 제대로 렌더링 되고 있는지 확인
- Card 컴포넌트에서 데이터를 직접 import해 활용하던 방식을 수정하면서 `setData` 메서드를 `setProps`로 변경

## 특이사항

* [x] 구현하면서 궁금했던 점이나, 아쉬웠던 부분이 있으면 적어주세요.

> 개인적으로 이번 미션 전후로 컨디션이 안 좋았는데, 그 때문에 요구사항들을 꼼꼼히 숙지하지 못하고 급하게 과제를 제출했습니다.
피드백 주신 내용들을 바탕으로 부족했던 점들을 사후에나마 보완해보았고, 남은 기간 동안 컨디션 관리에도 신경을 써야겠습니다 🥲

## 제출 후 확인

* [x] 본인의 구현 결과를 `dev` 브랜치에 구현하고, `dev` 브랜치에서 `submission` 브랜치로 PR을 보내주세요.
* [x] Pull Request 생성 후 Reivewer에 @externship-master 와 @lkaybob 을 추가해주세요.

jambottle committed:
Test: 3주차 미션 요구사항에 대한 컴포넌트 단위 테스트 코드 작성
- Jest와 Vue Test Utils를 활용하여 다섯 가지 미션 요구사항에 대한 테스트 코드 작성

jambottle committed:
Add: 3주차 미션 요구사항 구현 - (1) Card 컴포넌트 (Card.vue)
- Header 영역과 Nav 영역 사이에 2열로 표시될 상품 정보 카드의 레이아웃 구현 및 스타일링
- 상품 정보 카드는 상품의 사진과 이름, 가격을 포함하고, 할인율을 조건부 렌더링으로 표시

jambottle committed:
Add: 3주차 미션 요구사항 구현 - (2) ItemList 페이지 (ItemList.vue)
- 페이지 최상단과 최하단에 쇼핑몰 이름과 메뉴를 포함하는 Header 영역과 Nav 영역을 각각 구현
- 상품 정보 카드는 임의로 8개를 반복하고, Nav 영역은 스크롤과 상관없이 화면의 최하단에 고정

allegoryj commented:
재원님 안녕하세요!
상품 정보 카드를 8개를 반복해서 구현하신 점이 인상깊었습니다.
mock data를 불러서 사용하셔서 코드도 군더더기 없이 깔끔하네요 :)
저는 header랑 nav도 컴포넌트로 나누어서 구현했는데 볼륨이 작아서 굳이 나눌 필요가 없다고 생각하신걸까요?
재원님의 생각은 어떤지도 궁금합니다! 
 3주차 미션 고생많으셨습니다.

allegoryj commented on mission/src/views/ItemList.vue:
```diff
@@ -1,12 +1,134 @@
+    </header>
+
+    <section class="item-list-body">
+      <ItemListCard v-for="card in 8" :key="card" />
```
이런식으로 8개를 반복하신거군요!

jambottle replied:
네, 2열로 정렬이 잘 되는지 간단히 확인해볼 때에는 저렇게 v-for 디렉티브를 활용할 수 있겠더라고요-!

allegoryj commented on mission/src/views/ItemList.vue:
```diff
@@ -1,12 +1,134 @@
+  faBookmark,
+  faShoppingCart,
+  faUser,
+} from '@fortawesome/free-solid-svg-icons';
```
아이콘들은 이런식으로 가져올 수도 있군요.. 배워갑니다!

KyubumJang commented:
안녕하세요! 재원님!
상품 정보 카드를 별도록 구분하여서 구현하셨군요!
v-for을 사용해서 더 깔끔하게 코드를 구현하신 것 같습니다.
상세한 코멘트는 추후에 추가적으로 달아두도록 하겠습니다. 이번 주 고생 많으셨습니다!

Soohyeon02 commented:
안녕하세요 재원님!
Code Presentation 잘 보았습니다!
이번에도 역시 잘 구현하신 것 같습니다!
저도 v-for 과 컴포넌트를 사용해 보아야 할텐데 아직은 어렵네요,,,
미션 수행하느라 수고 많으셨습니다!

ddukddak-kim commented:
안녕하세요 재원님 :)
코드를 보니 간단하게 구현하신 듯 합니다. css를 잘 다루시네요!👍👍
3주차 미션 수행하느라 고생 많으셨고 4주차도 화이팅 해보아욥 ☺

watertheqlant commented on mission/src/components/ItemList/Card.vue:
```diff
@@ -0,0 +1,94 @@
+  .item-desc {
+    width: 180px;
+    overflow: hidden;
+    text-overflow: ellipsis;
+    white-space: nowrap;
```
js 에서 비슷한 로직을 구현했는데 css 에서 처리도 가능했군요.
😎 감탄 했습니다

jambottle replied:
JS로는 어떻게 구현하셨는지 저도 한번 보러 가야겠네요 😃

watertheqlant commented on mission/src/views/ItemList.vue:
```diff
@@ -1,12 +1,134 @@
+      }
+    }
+  }
+}
```
정말.. 중복없이 깔끔한 style 코드 이네요. 
보면서 계속 박수치게 됩니다. 🤗

watertheqlant commented on mission/tests/unit/ItemList/Card.spec.js:
```diff
@@ -0,0 +1,28 @@
+    const testItemDiscountPrice = 4000000000;
+    const testItemDiscountRate = 1 - (testItemDiscountPrice / testItemOriginalPrice);
+
+    await wrapper.setData({
```
setData 해줄 때도 동기적 표현이 필요한가요? 🤔❔

jambottle replied:
질문이 이해가 잘 안 됐어요... 혹시 어떤 부분을 **동기적 표현**이라고 말씀하신 걸까요??

lkaybob commented on mission/src/components/ItemList/Card.vue:
```diff
@@ -0,0 +1,94 @@
+export default {
+  name: 'ItemListCard',
+
+  data() {
```
이번 주차의 의도는 props를 활용한 list rendering을 적용하는 것이었습니다. 
* 때문에 상품 목록에 대한 정보를 list(ItemList.vue)로 받아온다고 했을 때 개별 상품 정보(Card.vue)를 prop으로 받아와서 구현을 하는 것이 필요했습니다. (data로 받아오는 것이 아님)

jambottle replied:
피드백 받은 대로 상위 컴포넌트에서 전달한 props가 리스트 렌더링에 적용될 수 있도록 수정해보았습니다! ⇒ **[결과](https://github.com/jambottle/ably-externship-frontend/pull/3/files#diff-c6e4d5c0a06b1cb3e793f08e8ffb6b362e00569e8af662903dcbc2b9fc8e3490)**

아무래도 지난 주 미션 때 만든 Mock Data가 있다 보니 해당 데이터를 재사용하는 걸 우선으로 생각했던 것 같습니다.
그런데, 결국에는 어디까지나 "Mock" Data이기 때문에 해당 데이터의 재사용을 굳이 우선시할 필요가 없음을 깨달았습니다 🤔

lkaybob commented on mission/tests/unit/ItemList/Item.spec.js:
```diff
@@ -1,10 +0,0 @@
-import { mount } from '@vue/test-utils';
```
테스트 코드 작성이 없네요

jambottle replied:
**테스트 코드 작성은 [ItemList.spec.js](https://github.com/jambottle/ably-externship-frontend/pull/3/files#diff-ce80fe4aba5830c9bd668bad16ab4ba9ecba2e43a0fbc600930207e51f7363eb)와 [Card.spec.js](https://github.com/jambottle/ably-externship-frontend/pull/3/files#diff-5d8cefe983f0b0e4c2813653b055fb43fe509ea9d95ef238cd18c82421b7617a) 두 파일에서 진행했습니다!**
혹시 이 부분에 대해서는 다시 한 번 확인해주실 수 있을까요? 😅

lkaybob replied:
제가 놓쳤었네요. 죄송합니다.
* 테스트 코드의 테스트케이스들이 단순한 점이 아쉽습니다. 요구사항을 세분화해서 구현해보시는 것이 더 이해하기 편할 듯 합니다.
* List의 경우에도 더 구현할 수 있는 테스트 케이스가 존재했는데 (예. 상품 목록에 따른 컴포넌트 렌더링 갯수 확인) 이 부분이 없어 아쉽습니다.

jambottle replied:
아닙니다, 바쁘신 와중에도 피드백 남겨주셔서 감사합니다! 🙏

개인적으로 지난 미션 전후로 컨디션이 안 좋았는데, 그 때문에 요구사항들을 꼼꼼히 숙지하지 못하고 급하게 과제를 제출했습니다.
오늘이나마 피드백 주신 내용들을 포함해 부족했던 점들을 보완해보면서 의도하신 학습 목표에 맞는 코드로 마무리 지어보겠습니다!!

lkaybob commented on mission/src/views/ItemList.vue:
```diff
@@ -1,12 +1,134 @@
+    </header>
+
+    <section class="item-list-body">
+      <ItemListCard v-for="card in 8" :key="card" />
```
Card.vue에서 남긴 comment와 동일합니다. Card.vue에서 prop을 정의하고 여기서 개별 item에 대한 정보를 넘겨줬어야 합니다.

jambottle replied:
_(수정 완료; 자세한 내용은 위에 남긴 코멘트와 동일합니다.)_

lkaybob reviewed (commented):
외관상 요구사항은 만족하나, 내부 구현은 이번 주차의 의도대로 구현해주지 못하셨습니다. 또한 테스트 코드 작성이 없는 부분이 매우 아쉽습니다. Inline comment를 확인해주세요

jambottle committed:
Modify: ItemList 컴포넌트로부터 Header/Footer 컴포넌트 분리
- Card 컴포넌트로 props를 전달하는 과정을 명확히 나타내기 위해 관련성이 낮은 부분을 분리

jambottle committed:
Modify: Mock Data 타입 변경 (itemListData.js)
- Card 컴포넌트의 리스트 렌더링에 활용하기 위해 Array 타입의 Mock Data 새로 작성

jambottle committed:
Modify: Card 컴포넌트의 데이터 import 방식 변경
- 학습 목표 및 요구사항에 따라 Card 컴포넌트에서 데이터를 직접 import하는 방식 지양
- ItemList 페이지에서 import한 데이터를 props로 전달받아 데이터 바인딩 및 렌더링 구현

jambottle committed:
Modify: Card 컴포넌트의 props 전달 방식 변경
- Card 컴포넌트로 props를 전달할 때, 데이터를 통째로 넘기는 것 대신 분해하여 전달
- Card 컴포넌트에서 props를 받을 때, prop 별 데이터 타입 및 기본값(default) 지정
- 위의 변경사항에 따라 컴포넌트 내에서의 데이터 바인딩 방식 조정 등

jambottle committed:
Add: 하단 스크롤 여부에 따라 Header 영역을 숨기는 효과 구현
- Method와 Lifecycle Hooks를 활용하여 하단 스크롤 여부를 감지
- 하단 스크롤 여부를 hasScrolled 변수에 Boolean 타입으로 저장
- 삼항 연산자로 Header 영역에 조건부 스타일을 부여하여 숨김 여부를 결정

jambottle committed:
Fix: Header 영역을 숨기는 효과 구현에 따라 레이아웃 재조정

jambottle committed:
Refactor: 테스트 코드 리팩토링 - (1) Header 컴포넌트에 대한 테스트 코드 분리

jambottle committed:
Refactor: 테스트 코드 리팩토링 - (2) Footer 컴포넌트에 대한 테스트 코드 분리

jambottle committed:
Refactor: 테스트 코드 리팩토링 - (3) Card 컴포넌트에 대한 기존 테스트 코드 수정
- Card 컴포넌트의 데이터 import 방식 변경에 따라 setData 메서드를 setProps로 변경
- 새로 작성한 Mock Data의 데이터 구조에 맞춰 테스트 코드에 활용할 데이터 구조 변경
- 기존 테스트 코드에서는 검증하지 않은 '상품 설명' 부분에 대한 테스트 케이스 추가

jambottle committed:
Refactor: 테스트 코드 리팩토링 - (4) ItemList 페이지에 대한 기존 테스트 코드 수정
- 리스트 렌더링에 활용될 임의의 데이터(Array 타입)를 setData 메서드로 테스트에 주입
- Card 컴포넌트가 리스트 렌더링으로 반복되는 횟수를 findAllComponents 메서드로 검증
- for 반복문을 활용해 컴포넌트별로 포함해야 할 요소들이 제대로 렌더링 되고 있는지 확인

jambottle commented:
**📌 테스트 코드 리팩토링 관련**

> 이번 주에는 학습 목표와 요구사항에 대한 파악이 부족했던 결과로 인해 만족할만한 수준으로 테스트 코드를 작성하지 못했습니다. 리팩토링을 통해 부족했던 점들을 최대한 보완해보았고, 덕분에 컴포넌트 간의 통신(특히 `Props`)에 대해서도 테스트 코드 검증이 가능하다는 것을 익힐 수 있었습니다.

**📌 느낀 점과 추가로 알게 된 점 관련**

> 개인적으로 이번 미션 전후로 컨디션이 안 좋았는데, 그 때문에 요구사항들을 꼼꼼히 숙지하지 못하고 급하게 과제를 제출했습니다.
피드백 주신 내용들을 바탕으로 부족했던 점들을 사후에나마 보완해보았고, 남은 기간 동안 컨디션 관리에도 신경을 써야겠습니다 🥲
